### PR TITLE
Cerus Combat Replay

### PR DESCRIPTION
**Pull request by Linkaaaaa**
## Cry of Rage
- [x] Normal
- [x] Empowered
- [x] CM Normal
- [x] CM Empowered

## Wail of Despair
- [x] Normal
- [x] Empowered
- [x] CM Normal
- [x] CM Empowered

## Pool of Despair
- [x] Normal
- [x] Empowered
- [x] CM Normal
- [x] CM Empowered

## Envious Gaze
- [x] Normal
- [x] Empowered
- [x] CM Normal
- [x] CM Empowered

## Crushing Regret
- [x] Normal
- [x] Empowered 
- [x] CM Normal
- [x] CM Empowered

## Malicious Intent
- [x] Normal
- [x] Empowered 
- [x] CM Normal
- [x] CM Empowered

## Insatiable Hunger
- Not Trackable i believe

## Enraged Smash
- [x] Normal
- [x] CM

## Extras
- Display invulnerable embodiments during split phase
- Renamed some skill id consts
- Added using ArcDPSEnums and removed it in the code
- Added Enraged Smash phase
- Adjusted all mechanics for challenge mode
- Added Dagda's Achievement Eligibilities (Danced with the Stars & Precision Anxiety)
- Added Cerus' Unbounded Optimism Achievement Eligibility
- Added Cerus' Apathetic eligibility with custom id until we know more
- Moved Malicious Shadow (NM & CM) to Targets for combat replay health display

### CODE REVIEW

File: GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs [hunk 1]
```diff
@@ -92,27 +100,30 @@ protected override List<int> GetTargetsIDs()
                 (int)TrashID.PermanentEmbodimentOfMalice,
                 (int)TrashID.PermanentEmbodimentOfRage,
                 (int)TrashID.PermanentEmbodimentOfRegret,
+                (int)TrashID.MaliciousShadow,
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs:103]: Does the normal mode shadow need to be in target?

Comment (Linkaaaaa) [reply GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs:103]: I just wanted it consistent with cm before people report us as bug "malicious shadow shows in cm but not nm" :D

File: GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs [hunk 2]
```diff
@@ -266,6 +285,9 @@ internal override void EIEvtcParse(ulong gw2Build, int evtcVersion, FightData fi
                             target.OverrideName("Empowered " + target.Character);
                         }
                         break;
+                    case (int)TrashID.MaliciousShadow:
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs:288]: Can be regrouped with the case below

File: GW2EIEvtcParser/EncounterLogic/Strikes/SotO/CosmicObservatory.cs [hunk 3]
```diff
@@ -482,5 +483,39 @@ internal override string GetLogicName(CombatData combatData, AgentData agentData
...
+                {
+                    IReadOnlyDictionary<long, BuffsGraphModel> bgms = player.GetBuffGraphs(log);
+                    if (bgms != null && bgms.TryGetValue(AchievementEligibilityPrecisionAnxiety, out BuffsGraphModel bgm))
+                    {
+                        if (bgm.BuffChart.Any(x => x.Value == 1))
+                        {
+                            buffCounter++;
+                        }
+                    }
+                    IReadOnlyList<DeathRecap> deaths = player.GetDeathRecaps(log);
```
[the remark this answers is not in the document]
Comment (Linkaaaaa) [reply GW2EIEvtcParser/EncounterLogic/Strikes/SotO/CosmicObservatory.cs:507]: on Player only GetDeathRecaps is available, i'll change to log.CombatData.GetDeadEvents(player.AgentItem)

File: GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs [hunk 4]
```diff
@@ -333,23 +522,214 @@ private static void AddHiddenWhileNotCasting(NPC target, ParsedEvtcLog log, Comb
...
+                        lifespan.end = Math.Min(lifespan.end, cast.Time);
+                    }
+                }
+            }
+            return lifespan;
+        }
+
+        private static (long start, long end) ComputeEmbodimentMechanicLifespanWithCancellationTime(ParsedEvtcLog log, (long start, long end) lifespan)
+        {
+            IReadOnlyList<PhaseData> phases = log.FightData.GetPhases(log).Where(x => !x.BreakbarPhase && x.CanBeSubPhase).ToList();
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs:699]: This has a requirement on ComputePhases option being on, checking Cerus losing its invulnerability would be more robust

File: GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs [hunk 5]
```diff
@@ -333,23 +522,214 @@ private static void AddHiddenWhileNotCasting(NPC target, ParsedEvtcLog log, Comb
...
+                        double degreedRotatedOpposite = (lifespanDamageOppositeCancelled.end - lifespanDamageOppositeCancelled.start) / millisecondsPerDegreeOpposite;
+                        var rotation3 = new AngleConnector(facing, (float)degreedRotatedOpposite);
+                        var rectangle3 = (RectangleDecoration)new RectangleDecoration(width, 100, lifespanDamageOppositeCancelled, Colors.Red, 0.2, oppositeAgentConnector).UsingRotationConnector(rotation3);
+                        replay.Decorations.Add(rectangle3);
+                    }
+                }
+            }
+        }
+
+        private static (long start, long end) ComputeMechanicLifespanWithCancellationTime(ParsedEvtcLog log, (long start, long end) lifespan)
```
Comment (EliphasNUIT) [inline GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs:680]: This API can already take Cerus as Input as it is always called after a target.IsSpecies(Cerus) check

Comment (EliphasNUIT) [reply GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs:680]: (as an AgentItem, you can then use log.FindActor API)

Comment (Linkaaaaa) [reply GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs:680]: I could merge the 2 into one and have the target.isSpecies(cerus) inside

Comment (EliphasNUIT) [reply GW2EIEvtcParser/EncounterLogic/Strikes/SotO/TempleOfFebe.cs:680]: Sounds good to me

**Comment by EliphasNUIT** [comment]
Thanks!

**Pull request merged**